pythonx/UltiSnips/store.py

```diff
@@ -0,0 +1,237 @@
+        try :
+            return getattr(self._dict, key)
+        except :
+            try :
+                return self[key]
+            except :
+                pass
+            raise
+
+    def __setattr__(self, key, val):
+        if key[0] == '_' :
+            return super().__setattr__(key, val)
+        else :
+            self[key] = val
+
+    def __setitem__(self, key, val):
+        key = str(key)
+        return self._dict.__setitem__(key, val)
+    def __delitem__(self, *args, **kwargs):
+        return self._dict.__delitem__(*args, **kwargs)
+    def __len__(self, *args, **kwargs):
+        return self._dict.__len__(*args, **kwargs)
+    def __iter__(self, *args, **kwargs):
+        return self._dict.__iter__(*args, **kwargs)
+    def __iter__(self, *args, **kwargs):
+        return self._dict.__iter__(*args, **kwargs)
+
+    def load(self):
+        pass
+
+    def save(self):
+        pass
+
+
+class BufferStore(Store):
+    """
+    A data store local to a buffer. Any python interpolation in a snippet running the same buffer
+    access the same instance through `snip.store.buffer`
+    """
+    def __init__(self):
+        self._dict = dict()
+
+    _stores = defaultdict(lambda : BufferStore())
+
+    @classmethod
+    def bufferStore(cls, buffer_number):
+        return cls._stores[buffer_number]
+
+
+    @classmethod
+    def currentBufferStore(cls):
+        """
+        Gets or create the store for the current buffer.
+        """
+        return cls.bufferStore(vim_helper.buf.number)
+
+    @classmethod
+    def clean(cls):
+        """
+        Check each buffer if it is alive, else delete it.
+        """
+        keys = set(cls._stores.keys())
+        alive = { b.number for b in vim_helper.vim.buffers if b.valid }
+        for k in keys - alive:
+            del cls._stores[k]
+
+
+class SessionStore(Store):
+    """
+    A Store common to a vim instance. Any python interpolation of any snippet in any buffer can access
+    the same instance through `snip.store.session`
+    """
+    def __init__(self):
+        self._dict = dict()
+
+    _store = None
+    @classmethod
+    def get(cls):
+        if cls._store is None :
+            cls._store = cls()
+        return cls._store
+
+
+class _PersistentStore(Store):
+    """
+    A Store persistent saved to json format.
+    """
+    def __init__(self, path):
+        self._path = path
+        self.load()
+        self._changed = False
+
+    def load(self):
+        try :
+            with open(self._path, 'r') as f :
+                self._dict = json.load(f)
+        except :
+            self._dict = dict()
+
+    def save(self):
+        if self._changed :
+            with open(self._path, 'w') as f :
+                json.dump(self._dict, f)
+
+    def __setitem__(self, key, val):
+        self._changed = True
+        super().__setitem__(key, val)
+
+
+def _encodeFilePath(path):
+    return str(Path(path).resolve()).replace('%', '_%%_').replace('/', '%')
```

Comment:
what are the `%` doing? This seems to be unix focused, will this work on Windows as expected?

Reply:
The `.replace('/', '%')` encodes the "/" characters to avoid creating subdirs (in a similar way vim does for the swap files). Since it's not a reserved name character in neither posix nor windows, it should work (except may be the file name length limit, that may need specific testing with FAT32... I may look at how vim handles it with swap files).

the `.replace('%', '_%%_')` simply escapes the '%' character to make sure any file (even containing '%') has a unique file path. The pattern is complex to handle cases like "/path%/tofile". I don't really know why one would put "%" in a file or dir name... but you know, "always consider the user dumber than the computer".

Reply:
That is fragile, at windows you would need to replace '\\' as path separator and maybe other characters. How about you just sha1 hash the resolved path, you will not have any problems with special characters and a flat hierarchy with no collisions like this, however you cannot direct back from hash to filename. Which might be considered a slight security plus.

Alternatively you could use base32 to encode bijectively between paths and save filenamen (see https://stackoverflow.com/questions/3945541/base64-encoding-safe-for-filenames).